### PR TITLE
Jamesb issue283

### PR DESCRIPTION
**Pull request by jbrowne6**
This new check for equivalence does not appear to effect performance, neither training time or accuracy.  The asserts are no longer triggered.

### CODE REVIEW

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h [hunk 1]
```diff
@@ -38,11 +38,11 @@ namespace fp{
 				}
 
 				inline bool checkInequality(const zipClassAndValue<int,double>& otherData){
-					return std::abs(featureVal - otherData.featureVal) > fpSingleton::getSingleton().returnDoubleEpsilon();
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:41]: seems like we should get rid of `returnFloatEpsilon()` and `returnDoubleEpsilon()` and their related variables in fpinfo and fpsingleton now

File: packedForest/src/fpSingleton/fpInfo.h [hunk 2]
```diff
@@ -78,8 +78,8 @@ namespace fp {
 				fractionOfFeaturesToTest=-1.0; 
 				binSize=0;
 				numCores=1;
-				double_epsilon=std::numeric_limits<double>::epsilon()*10;
-				float_epsilon=std::numeric_limits<float>::epsilon()*10;
+				double_epsilon=std::numeric_limits<double>::epsilon();
```
Comment (falkben) [inline packedForest/src/fpSingleton/fpInfo.h]: is `double_epsilon`/`double_float` needed?  doesn't appear to be called anywhere

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h [hunk 3]
```diff
@@ -38,11 +38,11 @@ namespace fp{
...
+					return std::abs(featureVal - otherData.featureVal) > std::numeric_limits<double>::epsilon() * 4 * std::abs(featureVal + otherData.featureVal);
 				}
 
 				inline bool checkInequality(const zipClassAndValue<int,float>& otherData){
-					return std::abs(featureVal - otherData.featureVal) > fpSingleton::getSingleton().returnFloatEpsilon();
+					return std::abs(featureVal - otherData.featureVal) > std::numeric_limits<float>::epsilon() * 4 * std::abs(featureVal + otherData.featureVal);
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:45]: i'm missing the intuition here about why we want to add the two `featureVal`s together when checking the inequality

**Review by falkben — CHANGES_REQUESTED** [review]
Thanks very much for this.  However, I am not sure we're doing the floating point check correctly.

This stack overflow post has some good discussion. https://stackoverflow.com/questions/17333/what-is-the-most-effective-way-for-float-and-double-comparison

Specifically, epsilon changes with the scale of your number (true epsilon is only valid for 1.0).  For now, it seems you are handling it with epsilon*4, but I think this would break with big numbers.  I also don't see how adding the numbers together on the right side of the inequality check returns what we want.

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h [hunk 4]
```diff
@@ -38,11 +38,11 @@ namespace fp{
 				}
 
 				inline bool checkInequality(const zipClassAndValue<int,double>& otherData){
-					return std::abs(featureVal - otherData.featureVal) > fpSingleton::getSingleton().returnDoubleEpsilon();
+					return std::abs(featureVal - otherData.featureVal) > std::numeric_limits<double>::epsilon() * 4 * std::abs(featureVal + otherData.featureVal);
```
Comment (falkben) [inline packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:41]: we might also want `fabs` here...?  it always returns a float...

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h (hunk 1 above)
Comment (jbrowne6) [reply packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:41]: agreed.  Im removing them.

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h (hunk 4 above)
Comment (jbrowne6) [reply packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:41]: fabs == abs in c++.  Only different in C.

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h (hunk 3 above)
Comment (jbrowne6) [reply packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:45]: I took this from the c++ reference.  https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon

File: packedForest/src/fpSingleton/fpInfo.h (hunk 2 above)
Comment (jbrowne6) [reply packedForest/src/fpSingleton/fpInfo.h]: No longer needed.  They are being removed.

File: packedForest/src/forestTypes/binnedTree/zipClassAndValue.h (hunk 3 above)
Comment (jbrowne6) [reply packedForest/src/forestTypes/binnedTree/zipClassAndValue.h:45]: It does seem like more work than is necessary.  I will test the solution you sent.

**Review by falkben — APPROVED** [review]

**Comment by MrAE** [comment]
@jbrowne6 good to merge this in?

**Comment by jbrowne6** [comment]
Please do.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Pull request merged**